src/clientapi/browser/launchClient.js

```diff
@@ -14,282 +14,143 @@
-
-if (DEBUG) {
-	import ..debugging._DEBUG as _DEBUG;
-	GLOBAL._DEBUG = new _DEBUG();
-
-
-	var DEVICE_ID_KEY = '.devkit.deviceId';
-	var deviceId;
-	var deviceType;
-
-	if (isSimulator) {
-		deviceId = CONFIG.simulator.deviceId;
-		deviceType = CONFIG.simulator.deviceType;
-
-		// simulate device chrome, input, and userAgent
-		var deviceType = uri.hash('deviceType');
-		if (deviceType) {
-			// hack to access SDK static resolutions file from a debug device
-			try {
-				import .simulateDevice;
-				var xhr = new XMLHttpRequest();
-				xhr.open('GET', '/simulate/static/util/resolutions.js', false);
-				xhr.send();
-
-				var resolutions = eval("(function () { exports = {}; " + xhr.responseText + "; return exports; })()");
-				simulateDevice.simulate(resolutions.get(deviceType));
-			} catch (e) {
-				logger.error(e);
-			}
-		}
-
-		// TODO: debugging conn should be fixed (missing socketio.js file)
-		// and should be changed to handle the load failure
-		debugging.conn.connect({
-			handshake: {
-				deviceId: deviceId,
-				deviceType: deviceType,
-				userAgent: navigator.userAgent,
-				screen: {
-					width: device.screen.width,
-					height: device.screen.height
-				}
-			}
-		}, function () { startApp(); });
-
-	} else {
-		deviceId = localStorage.getItem(DEVICE_ID_KEY);
-		if (!deviceId) {
-			import std.uuid;
-			deviceId = std.uuid.uuid();
-			localStorage.setItem(DEVICE_ID_KEY, deviceId);
-		}
-
-		if (device.isAndroid) {
-			deviceType = 'browser-android';
-		} else if (device.isIOS) {
-			deviceType = 'browser-ios';
-		} else {
-			deviceType = 'browser-mobile';
-		}
-
-		// start app without debugging connection
-		queueStart();
-	}
-
+CONFIG.isMuted = mute !== undefined && mute !== 'false' && mute !== '0' && mute !== 'no';
+
+var simulatorModules;
+if (DEBUG && isSimulator) {
+  simulatorModules = [];
+
+  // client API inside simulator: call onLaunch() on each simulator module,
+  // optionally block on a returned promise for up to 5 seconds
+  Promise
+    .map(CONFIG.simulator.modules, function (name) {
+      try {
+        var module = jsio(name);
+        if (module) {
+          simulatorModules.push(module);
+          if (typeof module.onLaunch == 'function') {
+            return module.onLaunch();
+          }
+        }
+      } catch (e) {}
+    })
+    .timeout(5000)
+    .finally(queueStart);
 } else {
-	queueStart();
+  queueStart();
 }
 
 function queueStart() {
-	if (window.GC_LIVE_EDIT && GC_LIVE_EDIT._isLiveEdit) {
-		var intervalId = setInterval(function(){
-			if (GC_LIVE_EDIT._liveEditReady) {
-				try {
-					startApp();
-				} catch(err) {
-					// In case loading fails, we will still clear the interval
-					console.error('Error while starting app', err);
-				}
-				clearInterval(intervalId);
-			}
-		}, 100);
-	} else {
-		startApp();
-	}
+  /* jshint -W117 */
+  if (window.GC_LIVE_EDIT && GC_LIVE_EDIT._isLiveEdit) {
+    var intervalId = setInterval(function(){
```

Comment:
Not that it was added in this PR, but this should probably be reworked. It seems like it is waiting for some async thing to happen, but it just polls instead of using callback/promise or similarly.